### ISSUE TITLE
Mutating readonly struct using embedded struct and pointers

### ISSUE THREAD

**Issue by SergeyTeplyakov** [dotnet/roslyn#26220]
**Version Used**:  master

The C# compiler legitimately complains about the following code:
```csharp
public unsafe struct FixedBytes
{
    // error CS0106: The modifier 'readonly' is not valid for this item
    private readonly fixed byte _bytes[8];
}
```

Which is fine, because the compiler can not prove that there would be potential mutations of the `_bytes` field in 'fixed' contexts.

But instead of using 'fixed' declaration, we can create a custom struct with the required size:

```csharp
public unsafe readonly struct FixedBytes
{
    [StructLayout(LayoutKind.Sequential, Size = 8)]
    private readonly struct FixedBuffer
    {
        public readonly byte FixedElementField;
    }

    private readonly FixedBuffer _bytes;
}
```
This is actually great (even desirable for my current project), but it means that the two very similar features are a bit inconsistent and the second version is actually allows us to create a mutable readonly struct:

```csharp
public unsafe readonly struct FixedBytes
{
    [StructLayout(LayoutKind.Sequential, Size = 8)]
    private readonly struct FixedBuffer
    {
        public readonly byte FixedElementField;
    }

    private readonly FixedBuffer _bytes;

    public byte this[int index]
    {
        get
        {
            fixed (byte* p = &_bytes.FixedElementField)
            {
                return p[index];
            }
        }
        set
        {
            fixed (byte* p = &_bytes.FixedElementField)
            {
                p[index] = value;
            }
        }
    }
}
```

And now you can see that the instance is mutable even though the struct declared as readonly:

```csharp
var ro = new FixedBytes();
ro[0] = 42;
Console.WriteLine(ro[0]); // prints 42
```

To be clear, I don't mind this hole. As I've mentioned above, I'm really glad that it exists (I'm going to use the version without the setter, so I'm not going to mutate readonly struct anyway).

**But I'm not quite sure is it a bug or the behavior for the second case is an expected one.** Can I expect this behavior to be the same in the future versions of the C# compiler?

**Comment by ufcpp** [comment]
Not only fixed buffers but also unsafe features allow readonly structs to be mutated.

```cs
readonly struct A
{
    public readonly int X;
    public A(int x) => X = x;
}

class Program
{
    unsafe static void Main()
    {
        A x = new A(1);
        int* p = (int*)&x;
        *p = 2;
        System.Console.WriteLine(x.X); // 2
    }
}
```

**Comment by SergeyTeplyakov** [comment]
@ufcpp I can see why the compiler does not warn in this case as well as in other cases (like with c-like unions based on explicit field offset). But the compiler definitely can detect that the '&' operator is applied to a readonly field and warn/fail about it.

Actually, R# emits an error in this case and I thought that the code that I provided is not correct before I've decided to compile it.

![image](https://user-images.githubusercontent.com/726448/38911391-d03529e4-4283-11e8-95e8-98e05fb37745.png)
